gc/gc-iceberg/src/main/java/org/projectnessie/gc/iceberg/ExpireContentsProcedure.java

```diff
@@ -0,0 +1,292 @@
+  private static final String COL_METADATA_LOCATION = "metadataLocation";
+  private static final String COL_SNAPSHOT_ID = "snapshotId";
+
+  // columns used for intermediate transforms
+  private static final String COL_EXPIRED_FILES_ARRAY = "expiredFilesArray";
+  private static final String COL_EXPIRED_FILES_WITH_TYPE = "expiredFilesWithType";
+  private static final String COL_EXPIRED_FILES_AND_TYPE = "expiredFilesAndType";
+  private static final String COL_EXPIRED_FILES = "expiredFiles";
+  private static final String COL_EXPIRED_FILES_TYPE = "expiredFilesType";
+  private static final String COL_EXPIRED_FILES_COUNT = "expiredFilesCount";
+  private static final String COL_EXPIRED_FILES_LIST = "expiredFilesList";
+
+  public ExpireContentsProcedure(TableCatalog currentCatalog) {
+    super(currentCatalog);
+  }
+
+  @Override
+  public ProcedureParameter[] parameters() {
+    return PARAMETERS;
+  }
+
+  @Override
+  public StructType outputType() {
+    return OUTPUT_TYPE;
+  }
+
+  @Override
+  public String description() {
+    return String.format(
+        "Expires the Iceberg snapshots that are collected by '%s' procedure.",
+        IdentifyExpiredContentsProcedure.PROCEDURE_NAME);
+  }
+
+  @Override
+  public InternalRow[] call(InternalRow internalRow) {
+    String gcCatalogName = internalRow.getString(0);
+    String gcOutputBranchName = internalRow.getString(1);
+    String gcOutputTableIdentifier = internalRow.getString(2);
+    Map<String, String> nessieClientConfig = new HashMap<>();
+    MapData map = internalRow.getMap(3);
+    for (int i = 0; i < map.numElements(); i++) {
+      nessieClientConfig.put(
+          map.keyArray().getUTF8String(i).toString(), map.valueArray().getUTF8String(i).toString());
+    }
+    String runId = !internalRow.isNullAt(4) ? internalRow.getString(4) : null;
+    boolean dryRun = !internalRow.isNullAt(5) && internalRow.getBoolean(5);
+
+    IdentifiedResultsRepo identifiedResultsRepo =
+        new IdentifiedResultsRepo(
+            spark(), gcCatalogName, gcOutputBranchName, gcOutputTableIdentifier);
+
+    if (runId == null) {
+      runId = getLatestCompletedRunID(gcOutputTableIdentifier, identifiedResultsRepo);
+    }
+
+    try (FileIO fileIO = getFileIO(nessieClientConfig)) {
+      Dataset<Row> expiredContentsDF = getExpiredContents(runId, identifiedResultsRepo, fileIO);
+      if (!dryRun) {
+        expiredContentsDF =
+            expiredContentsDF.map(
+                new DeleteFunction(fileIO), RowEncoder.apply(expiredContentsDF.schema()));
+      }
+      List<Row> rows = expiredContentsDF.collectAsList();
+      return rows.stream()
+          .map(
+              row ->
+                  GCProcedureUtil.internalRow(
+                      // "content_id", "deleted_files_type", "deleted_files_count",
+                      // "deleted_files_list"
+                      row.getString(0), row.getString(1), row.getInt(3), row.getList(2)))
+          .toArray(InternalRow[]::new);
```

Comment:
This will result in _all_ rows with _all_ expired files manifesting on the Java heap at once. See comment about heap pressure below. This one, together with the `.collectAsList()` above, however is worse, because it manifests _all_ expired files.

I'm not sure whether it's worth to return every deleted file to the user. I'd expect that giving users some hint about _how_ to access the list of deleted files should be enough. I.e. a table that users can access separately.

Reply:
ok. I will write the output to table.